src/lsptoolshost/commands.ts

```diff
@@ -103,73 +103,70 @@ async function completionComplexEdit(
-        const newRange = editor.document.validateRange(
-            new vscode.Range(
-                textEdit.range.start.line,
-                textEdit.range.start.character,
-                textEdit.range.end.line,
-                textEdit.range.end.character
-            )
-        );
+    const componentName = '[roslyn.client.completionComplexEdit]';
 
-        // HACK:
-        // ApplyEdit would fail the first time it's called when an item was committed with text modifying commit char (e.g. space, '(', etc.)
-        // so we retry a couple time here as a tempory workaround. We need to either figure our the reason of the failure, and/or try the
-        // approach of sending another edit request to server with updated document.
-        for (let i = 0; i < 3; i++) {
-            if (isSnippetString) {
-                editor.selection = new vscode.Selection(newRange.start, newRange.end);
-                success = await editor.insertSnippet(new vscode.SnippetString(textEdit.newText));
-            } else {
-                const edit = new vscode.WorkspaceEdit();
-                const newTextEdit = vscode.TextEdit.replace(newRange, textEdit.newText);
-                edit.set(editor.document.uri, [newTextEdit]);
-                success = await vscode.workspace.applyEdit(edit);
-
-                if (success && newOffset >= 0) {
-                    const newPosition = editor.document.positionAt(newOffset);
-                    editor.selections = [new vscode.Selection(newPosition, newPosition)];
-                }
-            }
+    // Find TextDocument, opening if needed.
+    const document = await vscode.workspace.openTextDocument(uriStr);
+    if (document === undefined) {
+        outputAndThrow(outputChannel, `${componentName} Can't open document with path: '${uriStr}'`);
+    }
 
-            if (success) {
-                break;
-            }
+    // Use editor if we need to deal with selection or snippets.
+    let editor: vscode.TextEditor | undefined = undefined;
+    if (isSnippetString || newOffset >= 0) {
+        editor = await vscode.window.showTextDocument(document);
+        if (editor == undefined) {
+            outputAndThrow(outputChannel, `${componentName} Editor unavailable for document with path: '${uriStr}'`);
         }
     }
 
-    if (!success) {
-        const componentName = '[roslyn.client.completionComplexEdit]';
-        const errorMessage = 'Failed to make a complex text edit for completion.';
-        outputChannel.show();
-        outputChannel.appendLine(`${componentName} ${errorMessage}`);
-
-        if (editor === undefined) {
-            outputChannel.appendLine(
-                `${componentName} Can't find visible document with uri.fsPath: '${uri.fsPath}' and uri.path: '${uri.path}'`
-            );
-
-            outputChannel.appendLine(`${componentName} URIs of all visible documents:`);
-            for (const visibleEditor of vscode.window.visibleTextEditors) {
```

Comment:
@genlu do you think we still need to keep this logging or can it be removed?

Reply:
I found logging useless. I can remove it you also see no value here.

Reply:
Those were added when we tried to figure out the uri issue that broke complex edit. If we haven't seen any such issues recently, then I'm fine removing them

Reply:
Logging removed

Reply:
we should keep at least the logging we had when we were not able to apply the edits.  throwing an error doesn't always bubble up to a visible location.

My original comment was asking Gen about the logging of the visible editor stuff that was also removed here.  It sounds like this part is OK to remove.

Switching it back to how you originally had it and it should be good to go in.